pkg/thing/thing.go

```diff
@@ -77,67 +77,101 @@ type connection interface {
 	attributes(tokenID string, content contentType, payload string, names []string) (reply []byte, err error)
 }
 
-// Thing represents an AM Thing identity
-type Thing struct {
+// Thing represents a device or a service with a digital identity in the ForgeRock Identity Platform.
+type Thing interface {
+
+	// RequestAccessToken requests an OAuth 2.0 access token for a thing. The provided scopes will be included in the token
+	// if they are configured in the thing's associated OAuth 2.0 Client in AM. If no scopes are provided then the token
+	// will include the default scopes configured in the OAuth 2.0 Client.
+	RequestAccessToken(scopes ...string) (response AccessTokenResponse, err error)
+
+	// RequestAttributes requests the attributes with the specified names associated with the thing's identity.
+	// If no names are specified then all the allowed attributes will be returned.
+	RequestAttributes(names ...string) (response AttributesResponse, err error)
+
+	// Session returns the current session for the Thing.
+	Session() Session
+}
+
+// Session holds session data
+type Session interface {
+
+	// Token returns the session token
+	Token() string
+
+	// HasRestrictedToken returns true if the session has a restricted token
+	HasRestrictedToken() bool
+
+	// SigningKey returns the signing key associated with a restricted SSO token
+	SigningKey() crypto.Signer
+
+	// Nonce returns the session nonce
+	Nonce() int
+
+	// IncrementNonce increments the session nonce
+	IncrementNonce()
+
+	// Valid returns true if the session is valid
+	Valid() (bool, error)
+
+	// Reauthenticate the thing to create a new session
+	Reauthenticate() (session Session, err error)
+
+	// Logout the session
+	Logout() error
+}
+
+type defaultThing struct {
 	connection connection
 	handlers   []callback.Handler
-	session    *Session
+	session    Session
 }
 
-// Session holds session data
-type Session struct {
-	thing *Thing
+type defaultSession struct {
+	thing Thing
```

Comment:
I would make the `Thing` a `defaultThing` since we make the assumption that it is a `dafaultThing` in the struct methods